relay.go

```diff
@@ -694,3 +767,87 @@ func validateRelayMaxTimeout(d time.Duration, logger Logger) time.Duration {
 	).Warn("Configured RelayMaxTimeout is invalid, using default instead.")
 	return _defaultRelayMaxTimeout
 }
+
+type relayFragmentSender struct {
+	callReq            *lazyCallReq
+	framePool          FramePool
+	frameReceiver      frameReceiver
+	failRelayItemFunc  func(items *relayItems, id uint32, failure string)
+	outboundRelayItems *relayItems
+	origID             uint32
+	sentBytes          uint
+	call               RelayCall
+}
+
+func (r *Relayer) newFragmentSender(dstRelay frameReceiver, cr *lazyCallReq, origID uint32, call RelayCall) *relayFragmentSender {
+	// TODO(cinchurge): pool fragment senders
+	return &relayFragmentSender{
+		callReq:            cr,
+		framePool:          r.conn.opts.FramePool,
+		frameReceiver:      dstRelay,
+		failRelayItemFunc:  r.failRelayItem,
+		outboundRelayItems: r.outbound,
+		origID:             origID,
+		sentBytes:          0,
+		call:               call,
+	}
+}
+
+func (rfs *relayFragmentSender) newFragment(initial bool, checksum Checksum) (*writableFragment, error) {
+	frame := rfs.framePool.Get()
+	frame.Header.ID = rfs.callReq.Header.ID
+	if initial {
+		frame.Header.messageType = messageTypeCallReq
+	} else {
+		frame.Header.messageType = messageTypeCallReqContinue
+	}
+
+	contents := typed.NewWriteBuffer(frame.Payload[:])
+
+	// flags:1
+	flagsRef := contents.DeferByte()
+	flagsRef.Update(rfs.callReq.Payload[_flagsIndex])
+
+	if initial {
+		// Copy all data before the checksum for the initial frame
+		contents.WriteBytes(rfs.callReq.Payload[_flagsIndex+1 : rfs.callReq.checksumTypeOffset])
+	}
+
+	// checksumType:1
+	contents.WriteSingleByte(byte(checksum.TypeCode()))
+
+	// checksum: checksum.Size()
+	checksumRef := contents.DeferBytes(checksum.Size())
+
+	if initial {
+		// arg1~1: write arg1 to the initial frame
+		contents.WriteUint16(uint16(len(rfs.callReq.method)))
+		contents.WriteBytes(rfs.callReq.method)
+		checksum.Add(rfs.callReq.method)
+	}
+
+	// TODO(cinchurge): pool writableFragment
+	return &writableFragment{
+		flagsRef:    flagsRef,
+		checksumRef: checksumRef,
+		checksum:    checksum,
+		contents:    contents,
+		frame:       frame,
+	}, contents.Err()
+}
+
+func (rfs *relayFragmentSender) flushFragment(wf *writableFragment) error {
+	wf.frame.Header.SetPayloadSize(uint16(wf.contents.BytesWritten()))
+	rfs.sentBytes += uint(wf.frame.Header.FrameSize())
+
+	sent, failure := rfs.frameReceiver.Receive(wf.frame, requestFrame)
+	if !sent {
+		rfs.failRelayItemFunc(rfs.outboundRelayItems, rfs.origID, failure)
+		return nil
+	}
+	return nil
+}
+
+func (rfs *relayFragmentSender) doneSending() {
+	rfs.call.SentBytes(rfs.sentBytes)
```

Comment:
why are we waiting for `doneSending` to increment sent bytes? it would likely be more accurate to track the sent bytes as we sent the frame, so as part of the `frameReceiver.Receive` I think? That way, we increment the counter at the point we're sending, rather than waiting till the end, so we'll have more accurate per-second counters

that would also avoid us changing from `uint16` to `uint` I think

Reply:
🤦 that was entirely a misunderstanding. somehow I kept thinking that `call.Sent()` sets a per-call value rather than increments a counter. will change.